Modules/posixmodule.c

```diff
@@ -7379,6 +7378,86 @@ os_openpty_impl(PyObject *module)
 #endif /* defined(HAVE_OPENPTY) || defined(HAVE__GETPTY) || defined(HAVE_DEV_PTMX) */
 
 
+#ifdef HAVE_SETSID
+#if defined(TIOCSCTTY) || defined(HAVE_TTYNAME)
+#define HAVE_FALLBACK_LOGIN_TTY 1
+#endif /* defined(TIOCSCTTY) || defined(HAVE_TTYNAME) */
+#endif /* HAVE_SETSID */
+
+#if defined(HAVE_LOGIN_TTY) || defined(HAVE_FALLBACK_LOGIN_TTY)
+/*[clinic input]
+os.login_tty
+
+    fd: fildes
+    /
+
+Prepare the tty of which fd is a file descriptor for a new login session.
+
+Make the calling process a session leader; make the tty the
+controlling tty, the stdin, the stdout, and the stderr of the
+calling process; close fd.
+[clinic start generated code]*/
+
+static PyObject *
+os_login_tty_impl(PyObject *module, int fd)
+/*[clinic end generated code: output=495a79911b4cc1bc input=5f298565099903a2]*/
+{
+#ifdef HAVE_LOGIN_TTY
+    if (login_tty(fd) == -1) {
+        return posix_error();
+    }
+#else /* defined(HAVE_FALLBACK_LOGIN_TTY) */
+    /* Establish a new session. */
+    if (setsid() == -1) {
+        return posix_error();
+    }
+
+    /* The tty becomes the controlling terminal. */
+#ifdef TIOCSCTTY
+    if (ioctl(fd, TIOCSCTTY, (char *)NULL) == -1) {
+        return posix_error();
+    }
+#else /* defined(HAVE_TTYNAME) */
+    /* Fallback method (archaic); from Advanced Programming in the UNIX(R)
+     * Environment, Third edition, 2013, Section 9.6 - Controlling Terminal:
+     * "Systems derived from UNIX System V allocate the controlling
+     * terminal for a session when the session leader opens the first
+     * terminal device that is not already associated with a session, as
+     * long as the call to open does not specify the O_NOCTTY flag." */
+    char *tmppath = ttyname(fd);
+    if (tmppath == NULL) {
+        return posix_error();
+    }
+
+#define CLOSE_IF_NOT_FD(otherfd) \
+    if (fd != otherfd) { \
+        close(otherfd); \
+    } \
+
+    CLOSE_IF_NOT_FD(0);
+    CLOSE_IF_NOT_FD(1);
+    CLOSE_IF_NOT_FD(2);
+
+    int tmpfd = open(tmppath, O_RDWR);
```

Comment:
add a comment as to why this is opened to check for an error and immediately closed.

Reply:
Thank you for the review. According to https://stackoverflow.com/questions/51593530/code-explanation-for-glibc-login-tty-function-openttyname-immediately-f, this approach is historical and was used in UNIX System V derived systems when TIOCSCTTY was not available. I will definitely add the comment as a part of the next commit with the other changes. It would be something like "explicitly open the tty to make it become a controlling tty" as in the fallback code for `fork()` in `pty.py` here: https://github.com/python/cpython/blob/main/Lib/pty.py